text/0162-refactoring-support.md

```diff
@@ -0,0 +1,702 @@
+CloudFormation will create a new resource, with the new logical ID, and possibly
+delete the old one. For stateful resources, this may cause interruption of
+service or data loss, or both.
+
+Historically, we have advised developers to avoid changing logical IDs. But this
+is sometimes impractical or conflicts with good software engineering practices.
+For instance, you may want to consolidate duplicated code across different CDK
+applications into a single reusable construct (an "L3 construct"). Introducing a
+new node for the L3 construct in the construct tree will rename the logical IDs
+of the resources in that subtree.
+
+Also, you might need to move resources within the tree for better readability,
+or between stacks to isolate concerns. Accidental renames have also caused
+issues for customers in the past. Perhaps even worse, if you depend on a
+third-party construct library, you are not in control of the logical IDs of
+those resources. If the library changes the logical IDs from one version to
+another, you will be affected without any action on your part.
+
+To address all these problems, the CDK CLI now automatically detects these
+cases, and, with your supervision, uses the new CloudFormation stack refactoring
+API to move the resources around.
+
+### How it works
+
+Suppose your CDK application has a single stack, called `MyStack`, containing an
+S3 bucket, a CloudFront distribution and a Lambda function. The construct tree
+looks like this (L1 constructs omitted for brevity):
+
+    App
+    └─ MyStack
+       ├─ Bucket
+       ├─ Distribution
+       └─ Function
+
+And suppose you make the following changes, after having deployed it to your AWS
+account:
+
+- Rename the bucket from `Bucket` to the more descriptive name `Origin`.
+- Move the bucket and the distribution under a new L3 construct called
+  `Website`, to make this pattern reusable in different applications.
+- Move the web-related constructs (now under the `Website` L3) to a new stack
+  called `Web`, for better separation of concerns.
+- Rename the original stack to `Service`, to better reflect its new specific
+  role in the application.
+
+The refactored construct tree looks like this:
+
+    App
+    ├─ Web
+    │  └─ Website
+    │     ├─ Origin
+    │     └─ Distribution
+    └─ Service
+       └─ Function
+
+Even though none of the resources have changed, their paths have (from
+`MyStack/Bucket/Resource` to `Web/Website/Origin/Resource` etc.) Since the CDK
+computes the logical IDs of the resources from their path in the tree, all three
+resources will have their logical IDs changed.
+
+From your perspective, the changes above are mere moves and renames, but what
+CloudFormation sees is the deletion of old resources and the creation of new
+ones. By opting in to the CDK refactoring support, the CDK CLI will work on your
+behalf to notify CloudFormation of your intention.
+
+To execute a refactor, run `cdk refactor --unstable=refactor`. The CLI will show
```

Comment:
what does the `--unstable` flag mean? and why do we name it `--unstable`?

Reply:
This is a generic flag we use in the CLI to indicate that the feature is unstable. By passing that flag, the user is acknowledging that.

Reply:
ahh I see, thank you for the clarification!